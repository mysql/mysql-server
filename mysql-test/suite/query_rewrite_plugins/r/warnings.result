#
# Test of rules where the parser gives warnings about the originial
# query, pattern or replacement.
#
call mtr.add_suppression("Plugin rewriter reported: 'Some rules failed to load.'");
# Query rewrite plugin was installed.
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'LOAD DATA INFILE "tab1.data" INTO TABLE test.t1',
'LOAD DATA INFILE "tab2.data" INTO TABLE test.t1' ),
( 'DO CONCAT("a", "b")', 'DO CONCAT("c", "d")' );
CALL query_rewrite.flush_rewrite_rules();
ERROR 45000: Loading of some rule(s) failed.
SELECT * FROM query_rewrite.rewrite_rules;
id	pattern	pattern_database	replacement	enabled	message
1	LOAD DATA INFILE "tab1.data" INTO TABLE test.t1	NULL	LOAD DATA INFILE "tab2.data" INTO TABLE test.t1	YES	Statement type of pattern not supported.
2	DO CONCAT("a", "b")	NULL	DO CONCAT("c", "d")	YES	Statement type of pattern not supported.
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SHOW /* I rewrote it */ WARNINGS' );
CALL query_rewrite.flush_rewrite_rules();
ERROR 45000: Loading of some rule(s) failed.
# We want the note about the query being rewritten here ...
SHOW WARNINGS;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
SHOW ERRORS;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
GET DIAGNOSTICS CONDITION 1 @p1 = RETURNED_SQLSTATE, @p2 = MESSAGE_TEXT;
SELECT @p1;
@p1
45000
SELECT @p2;
@p2
Loading of some rule(s) failed.
# ... but not here, since we failed parsing after seeing SHOW WARNINGS.
SHOW WARNINGS and give me a parse error, please;
ERROR 42000: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'and give me a parse error, please' at line 1
DELETE FROM query_rewrite.rewrite_rules;
# Test of warnings when we rewrite the diagnostics statement.  echo #
# We disable mtr's SHOW WARNINGS here, as it would just be confusing
# when it gets rewritten without us seeing it.
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SELECT "used to be SHOW WARNINGS"' ),
( 'SELECT 1', 'SHOW WARNINGS' );
CALL query_rewrite.flush_rewrite_rules();
ERROR 45000: Loading of some rule(s) failed.
SELECT 1;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
Note	1105	Query 'SELECT 1' rewritten to 'SHOW WARNINGS' by a query rewrite plugin
SHOW WARNINGS;
Level	Code	Message
Error	1644	Loading of some rule(s) failed.
Note	1105	Query 'SELECT 1' rewritten to 'SHOW WARNINGS' by a query rewrite plugin
Warnings:
Warning	1620	Plugin is busy and will be uninstalled on shutdown
# Query rewrite plugin was queued for uninstalling.
