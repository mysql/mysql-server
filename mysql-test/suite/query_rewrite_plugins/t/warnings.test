--echo #
--echo # Test of rules where the parser gives warnings about the originial
--echo # query, pattern or replacement.
--echo #

call mtr.add_suppression("Plugin rewriter reported: 'Some rules failed to load.'");

--source suite/query_rewrite_plugins/include/have_plugin_rewriter.inc
--source suite/query_rewrite_plugins/include/install_rewriter.inc

# Still many unsupported statements
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'LOAD DATA INFILE "tab1.data" INTO TABLE test.t1',
         'LOAD DATA INFILE "tab2.data" INTO TABLE test.t1' ),
       ( 'DO CONCAT("a", "b")', 'DO CONCAT("c", "d")' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();
SELECT * FROM query_rewrite.rewrite_rules;

INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SHOW /* I rewrote it */ WARNINGS' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();

--echo # We want the note about the query being rewritten here ...
SHOW WARNINGS;
SHOW ERRORS;
GET DIAGNOSTICS CONDITION 1 @p1 = RETURNED_SQLSTATE, @p2 = MESSAGE_TEXT;
SELECT @p1;
SELECT @p2;

--echo # ... but not here, since we failed parsing after seeing SHOW WARNINGS.
--error ER_PARSE_ERROR
SHOW WARNINGS and give me a parse error, please;

DELETE FROM query_rewrite.rewrite_rules;

--echo # Test of warnings when we rewrite the diagnostics statement.  echo #
--echo # We disable mtr's SHOW WARNINGS here, as it would just be confusing
--echo # when it gets rewritten without us seeing it.
--disable_warnings
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SELECT "used to be SHOW WARNINGS"' ),
       ( 'SELECT 1', 'SHOW WARNINGS' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();

SELECT 1;

SHOW WARNINGS;

--enable_warnings
--source suite/query_rewrite_plugins/include/uninstall_rewriter.inc
