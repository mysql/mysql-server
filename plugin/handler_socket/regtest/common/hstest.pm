
# vim:sw=2:ai

package hstest;

use DBI;
use Net::HandlerSocket;

our %conf = ();

sub get_conf_env {
  my ($key, $defval) = @_;
  return $ENV{$key} || $defval;
}

sub init_conf {
  $conf{host} = get_conf_env("MYHOST", "localhost");
  $conf{myport} = get_conf_env("MYPORT", 3306);
  $conf{dbname} = get_conf_env("MYDBNAME", "hstestdb");
  $conf{ssps} = get_conf_env("MYSSPS");
  $conf{user} = get_conf_env("MYSQLUSER", "root");
  $conf{pass} = get_conf_env("MYSQLPASS", "");
  $conf{hsport} = get_conf_env("HSPORT", 9998);
  $conf{hspass} = get_conf_env("HSPASS", undef);
}

sub get_dbi_connection {
  my ($dbname, $host, $myport, $ssps, $user, $pass)
    = ($conf{dbname}, $conf{host}, $conf{myport}, $conf{ssps},
      $conf{user}, $conf{pass});
  my $mycnf = "binary_my.cnf";
  my $dsn = "DBI:mysql:database=;host=$host;port=$myport"
    . ";mysql_server_prepare=$ssps"
    . ";mysql_read_default_group=perl"
    . ";mysql_read_default_file=../common/$mycnf";
  my $dbh = DBI->connect($dsn, $user, $pass, { RaiseError => 1 });
  return $dbh;
}

sub init_testdb {
  my $charset = $_[0] || "binary";
  my $dbh = get_dbi_connection();
  my $dbname = $conf{dbname};
  $dbh->do("drop database if exists $dbname");
  $dbh->do("create database $dbname default character set $charset");
  $dbh->do("use $dbname");
  return $dbh;
}

sub get_hs_connection {
  my ($host, $port) = @_;
  $host ||= $conf{host};
  $port ||= $conf{hsport};
  my $hsargs = { 'host' => $host, 'port' => $port };
  my $conn = new Net::HandlerSocket($hsargs);
  if (defined($conn) && defined($conf{hspass})) {
    $conn->auth($conf{hspass});
  }
  return $conn;
}


init_conf();

1;

