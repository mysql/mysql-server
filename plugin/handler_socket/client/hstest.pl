#!/usr/bin/perl

# vim:sw=8:ai:ts=8

use strict;
use warnings;

use DBI;
use Net::HandlerSocket;

my %conf = ();
for my $i (@ARGV) {
	my ($k, $v) = split(/=/, $i);
	$conf{$k} = $v;
}

my $verbose = get_conf("verbose", 0);
my $actions_str = get_conf("actions", "hsread");
my $tablesize = get_conf("tablesize", 10000);
my $db = get_conf("db", "hstest");
my $table = get_conf("table", "hstest_table1");
my $engine = get_conf("engine", "innodb");
my $host = get_conf("host", "localhost");
my $mysqlport = get_conf("mysqlport", 3306);
my $mysqluser = get_conf("mysqluser", "root");
my $mysqlpass = get_conf("mysqlpass", "");
my $hsport = get_conf("hsport", 9999);
my $loop = get_conf("loop", 10000);
my $op = get_conf("op", "=");
my $ssps = get_conf("ssps", 0);
my $num_moreflds = get_conf("moreflds", 0);
my $moreflds_prefix = get_conf("moreflds_prefix", "column0123456789_");
my $keytype = get_conf("keytype", "varchar(32)");
my $file = get_conf("file", undef);

my $dsn = "DBI:mysql:database=;host=$host;port=$mysqlport"
	. ";mysql_server_prepare=$ssps";
my $dbh = DBI->connect($dsn, $mysqluser, $mysqlpass, { RaiseError => 1 });
my $hsargs = { 'host' => $host, 'port' => $hsport };
my $cli = new Net::HandlerSocket($hsargs);

my @actions = split(/,/, $actions_str);
for my $action (@actions) {
	if ($action eq "table") {
		print("TABLE $db.$table\n");
		$dbh->do("drop database if exists $db");
		$dbh->do("create database $db");
		$dbh->do("use $db");
		my $moreflds = get_createtbl_moreflds_str();
		$dbh->do(
			"create table $table (" .
			"k $keytype primary key" .
			",v varchar(32) not null" .
			$moreflds .
			") character set utf8 collate utf8_bin " .
			"engine = $engine");
	} elsif ($action eq "insert") {
		print("INSERT $db.$table tablesize=$tablesize\n");
		$dbh->do("use $db");
		my $moreflds = get_insert_moreflds_str();
		for (my $i = 0; $i < $tablesize; $i += 100) {
			my $qstr = "insert into $db.$table values";
			for (my $j = 0; $j < 100; ++$j) {
				if ($j != 0) {
					$qstr .= ",";
				}
				my $k = "" . ($i + $j);
				my $v = "v" . int(rand(1000)) . ($i + $j);
				$qstr .= "('$k', '$v'";
				for (my $j = 0; $j < $num_moreflds; ++$j) {
					$qstr .= ",'$j'";
				}
				$qstr .= ")";
			}
			$dbh->do($qstr);
			print "$i/$tablesize\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "read") {
		print("READ $db.$table op=$op loop=$loop\n");
		$dbh->do("use $db");
		my $moreflds = get_select_moreflds_str();
		my $sth = $dbh->prepare(
			"select k,v$moreflds from $db.$table where k = ?");
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			# print "k=$k\n";
			$sth->execute($k);
			if ($verbose >= 10) {
				print "RET:";
				while (my $ref = $sth->fetchrow_arrayref()) {
					my $rk = $ref->[0];
					my $rv = $ref->[1];
					print " $rk $rv";
				}
				print "\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsinsert") {
		print("HSINSERT $db.$table tablesize=$tablesize\n");
		$cli->open_index(1, $db, $table, '', 'k,v');
		for (my $i = 0; $i < $tablesize; ++$i) {
			my $k = "" . $i;
			my $v = "v" . int(rand(1000)) . $i;
			my $r = $cli->execute_insert(1, [ $k, $v ]);
			if ($r->[0] != 0) {
				die;
			}
			print "$i/$tablesize\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsread") {
		print("HSREAD $db.$table op=$op loop=$loop\n");
		my $moreflds = get_select_moreflds_str();
		$cli->open_index(1, $db, $table, '', "k,v$moreflds");
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			# print "k=$k\n";
			my $r = $cli->execute_find(1, $op, [ $k ], 1, 0);
			if ($verbose >= 10) {
				my $len = scalar(@{$r});
				print "LEN=$len";
				for my $e (@{$r}) {
					print " [$e]";
				}
				print "\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsupdate") {
		my $vbase = "v" . int(rand(1000));
		print("HSUPDATE $db.$table op=$op loop=$loop vbase=$vbase\n");
		$cli->open_index(1, $db, $table, '', 'v');
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			my $v = $vbase . $i;
			print "k=$k v=$v\n";
			my $r = $cli->execute_update(1, $op, [ $k ], 1, 0,
				[ $v ]);
			if ($verbose >= 10) {
				print "UP k=$k v=$v\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsdelete") {
		print("HSDELETE $db.$table op=$op loop=$loop\n");
		$cli->open_index(1, $db, $table, '', '');
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			print "k=$k\n";
			my $r = $cli->execute_delete(1, $op, [ $k ], 1, 0);
			if ($verbose >= 10) {
				print "DEL k=$k\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "verify") {
		verify_do();
	}
}

sub verify_do {
	my ($fail_cnt, $ok_cnt) = (0, 0);
	my $sth = $dbh->prepare("select v from $db.$table where k = ?");
	use FileHandle;
	my $fh = new FileHandle($file, "r");
	while (my $line = <$fh>) {
		chomp($line);
		my @vec = split(/\t/, $line);
		my $k = $vec[3];
		my $v = $vec[7];
		next if (!defined($k) || !defined($v));
		# print "$k $v\n";
		$sth->execute($k);
		my $aref = $sth->fetchrow_arrayref();
		if (!defined($aref)) {
			print "FAILED: $k notfound\n";
			++$fail_cnt;
		} else {
			my $gv = $aref->[0];
			if ($gv ne $v) {
				print "FAILED: $k got=$gv expected=$v\n";
				++$fail_cnt;
			} else {
				print "OK: $k $v $gv\n" if $verbose >= 10;
				++$ok_cnt;
			}
		}
	}
	print "OK=$ok_cnt FAIL=$fail_cnt\n";
}

sub get_conf {
	my ($key, $def) = @_;
	my $val = $conf{$key};
	if ($val) {
		print "$key=$val\n";
	} else {
		$val = $def;
		$def ||= '';
		print "$key=$def(default)\n";
	}
	return $val;
}

sub get_createtbl_moreflds_str {
	my $s = "";
	for (my $j = 0; $j < $num_moreflds; ++$j) {
		$s .= ",$moreflds_prefix$j varchar(30)";
	}
	return $s;
}

sub get_select_moreflds_str {
	my $s = "";
	for (my $i = 0; $i < $num_moreflds; ++$i) {
		$s .= ",$moreflds_prefix$i";
	}
	return $s;
}

sub get_insert_moreflds_str {
	my $s = "";
	for (my $i = 0; $i < $num_moreflds; ++$i) {
		$s .= ",?";
	}
	return $s;
}

