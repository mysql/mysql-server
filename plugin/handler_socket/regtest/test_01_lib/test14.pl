#!/usr/bin/perl

# vim:sw=2:ai

# test for bugfix: commit/c88efe637f6a184b55d2bd8d060bda3e556572d8
# (some trailing bytes were dropped for varlen or nullable key fields)

BEGIN {
	push @INC, "../common/";
};

use strict;
use warnings;
use hstest;

my $dbh = hstest::init_testdb();
my $table = 'hstesttbl';
my $tablesize = 50;
$dbh->do(
  "create table $table (" .
  "k int primary key, " .
  "v1 varchar(30), " .
  "v2 varchar(30), " .
  "index i1(v1), index i2(v2, v1)) " .
  "engine = myisam default charset = binary");
srand(999);

my %valmap = ();

my $sth = $dbh->prepare("insert into $table values (?,?,?)");
for (my $i = 0; $i < $tablesize; ++$i) {
  my $k = $i;
  my ($s1, $s2) = ("", "");
  for (my $j = 0; $j < $i; ++$j) {
    $s1 .= chr(48 + $j % 10);
    $s2 .= chr(65 + $j % 10);
  }
  my $v1 = $s1;
  my $v2 = $s2;
  $sth->execute($k, $v1, $v2);
  $valmap{$k} = [ $v1, $v2 ];
}

dump_table();

my $hs = hstest::get_hs_connection(undef, 9999);
my $dbname = $hstest::conf{dbname};
$hs->open_index(1, $dbname, $table, '', 'k,v1,v2');
$hs->open_index(2, $dbname, $table, 'i1', 'k,v1,v2');
$hs->open_index(3, $dbname, $table, 'i2', 'k,v1,v2');

for (my $i = 0; $i <= 30; ++$i) {
  my ($v1, $v2) = @{$valmap{$i}};
  my ($rk, $rv1, $rv2);
  my $r = $hs->execute_single(1, '=', [ $i ], 1, 0);
  shift(@$r);
  ($rk, $rv1, $rv2) = @$r;
  print "PK $rk $rv1 $rv2\n";
  $r = $hs->execute_single(2, '=', [ $v1 ], 1, 0);
  shift(@$r);
  ($rk, $rv1, $rv2) = @$r;
  print "I1 $rk $rv1 $rv2\n";
  $r = $hs->execute_single(3, '=', [ $v2, $v1 ], 1, 0);
  shift(@$r);
  ($rk, $rv1, $rv2) = @$r;
  print "I2 $rk $rv1 $rv2\n";
}

sub dump_table {
  print "DUMP_TABLE\n";
  my $aref = $dbh->selectall_arrayref("select k,v1,v2 from $table order by k");
  for my $row (@$aref) {
    my ($k, $v1, $v2) = @$row;
    $v1 = "[null]" if !defined($v1);
    $v2 = "[null]" if !defined($v2);
    print "$k $v1 $v2\n";
    # print "MISMATCH\n" if ($valmap{$k} ne $v);
  }
}

