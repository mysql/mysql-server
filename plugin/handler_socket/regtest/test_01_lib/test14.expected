DUMP_TABLE
0  
1 0 A
2 01 AB
3 012 ABC
4 0123 ABCD
5 01234 ABCDE
6 012345 ABCDEF
7 0123456 ABCDEFG
8 01234567 ABCDEFGH
9 012345678 ABCDEFGHI
10 0123456789 ABCDEFGHIJ
11 01234567890 ABCDEFGHIJA
12 012345678901 ABCDEFGHIJAB
13 0123456789012 ABCDEFGHIJABC
14 01234567890123 ABCDEFGHIJABCD
15 012345678901234 ABCDEFGHIJABCDE
16 0123456789012345 ABCDEFGHIJABCDEF
17 01234567890123456 ABCDEFGHIJABCDEFG
18 012345678901234567 ABCDEFGHIJABCDEFGH
19 0123456789012345678 ABCDEFGHIJABCDEFGHI
20 01234567890123456789 ABCDEFGHIJABCDEFGHIJ
21 012345678901234567890 ABCDEFGHIJABCDEFGHIJA
22 0123456789012345678901 ABCDEFGHIJABCDEFGHIJAB
23 01234567890123456789012 ABCDEFGHIJABCDEFGHIJABC
24 012345678901234567890123 ABCDEFGHIJABCDEFGHIJABCD
25 0123456789012345678901234 ABCDEFGHIJABCDEFGHIJABCDE
26 01234567890123456789012345 ABCDEFGHIJABCDEFGHIJABCDEF
27 012345678901234567890123456 ABCDEFGHIJABCDEFGHIJABCDEFG
28 0123456789012345678901234567 ABCDEFGHIJABCDEFGHIJABCDEFGH
29 01234567890123456789012345678 ABCDEFGHIJABCDEFGHIJABCDEFGHI
30 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
31 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
32 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
33 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
34 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
35 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
36 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
37 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
38 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
39 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
40 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
41 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
42 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
43 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
44 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
45 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
46 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
47 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
48 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
49 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
PK 0  
I1 0  
I2 0  
PK 1 0 A
I1 1 0 A
I2 1 0 A
PK 2 01 AB
I1 2 01 AB
I2 2 01 AB
PK 3 012 ABC
I1 3 012 ABC
I2 3 012 ABC
PK 4 0123 ABCD
I1 4 0123 ABCD
I2 4 0123 ABCD
PK 5 01234 ABCDE
I1 5 01234 ABCDE
I2 5 01234 ABCDE
PK 6 012345 ABCDEF
I1 6 012345 ABCDEF
I2 6 012345 ABCDEF
PK 7 0123456 ABCDEFG
I1 7 0123456 ABCDEFG
I2 7 0123456 ABCDEFG
PK 8 01234567 ABCDEFGH
I1 8 01234567 ABCDEFGH
I2 8 01234567 ABCDEFGH
PK 9 012345678 ABCDEFGHI
I1 9 012345678 ABCDEFGHI
I2 9 012345678 ABCDEFGHI
PK 10 0123456789 ABCDEFGHIJ
I1 10 0123456789 ABCDEFGHIJ
I2 10 0123456789 ABCDEFGHIJ
PK 11 01234567890 ABCDEFGHIJA
I1 11 01234567890 ABCDEFGHIJA
I2 11 01234567890 ABCDEFGHIJA
PK 12 012345678901 ABCDEFGHIJAB
I1 12 012345678901 ABCDEFGHIJAB
I2 12 012345678901 ABCDEFGHIJAB
PK 13 0123456789012 ABCDEFGHIJABC
I1 13 0123456789012 ABCDEFGHIJABC
I2 13 0123456789012 ABCDEFGHIJABC
PK 14 01234567890123 ABCDEFGHIJABCD
I1 14 01234567890123 ABCDEFGHIJABCD
I2 14 01234567890123 ABCDEFGHIJABCD
PK 15 012345678901234 ABCDEFGHIJABCDE
I1 15 012345678901234 ABCDEFGHIJABCDE
I2 15 012345678901234 ABCDEFGHIJABCDE
PK 16 0123456789012345 ABCDEFGHIJABCDEF
I1 16 0123456789012345 ABCDEFGHIJABCDEF
I2 16 0123456789012345 ABCDEFGHIJABCDEF
PK 17 01234567890123456 ABCDEFGHIJABCDEFG
I1 17 01234567890123456 ABCDEFGHIJABCDEFG
I2 17 01234567890123456 ABCDEFGHIJABCDEFG
PK 18 012345678901234567 ABCDEFGHIJABCDEFGH
I1 18 012345678901234567 ABCDEFGHIJABCDEFGH
I2 18 012345678901234567 ABCDEFGHIJABCDEFGH
PK 19 0123456789012345678 ABCDEFGHIJABCDEFGHI
I1 19 0123456789012345678 ABCDEFGHIJABCDEFGHI
I2 19 0123456789012345678 ABCDEFGHIJABCDEFGHI
PK 20 01234567890123456789 ABCDEFGHIJABCDEFGHIJ
I1 20 01234567890123456789 ABCDEFGHIJABCDEFGHIJ
I2 20 01234567890123456789 ABCDEFGHIJABCDEFGHIJ
PK 21 012345678901234567890 ABCDEFGHIJABCDEFGHIJA
I1 21 012345678901234567890 ABCDEFGHIJABCDEFGHIJA
I2 21 012345678901234567890 ABCDEFGHIJABCDEFGHIJA
PK 22 0123456789012345678901 ABCDEFGHIJABCDEFGHIJAB
I1 22 0123456789012345678901 ABCDEFGHIJABCDEFGHIJAB
I2 22 0123456789012345678901 ABCDEFGHIJABCDEFGHIJAB
PK 23 01234567890123456789012 ABCDEFGHIJABCDEFGHIJABC
I1 23 01234567890123456789012 ABCDEFGHIJABCDEFGHIJABC
I2 23 01234567890123456789012 ABCDEFGHIJABCDEFGHIJABC
PK 24 012345678901234567890123 ABCDEFGHIJABCDEFGHIJABCD
I1 24 012345678901234567890123 ABCDEFGHIJABCDEFGHIJABCD
I2 24 012345678901234567890123 ABCDEFGHIJABCDEFGHIJABCD
PK 25 0123456789012345678901234 ABCDEFGHIJABCDEFGHIJABCDE
I1 25 0123456789012345678901234 ABCDEFGHIJABCDEFGHIJABCDE
I2 25 0123456789012345678901234 ABCDEFGHIJABCDEFGHIJABCDE
PK 26 01234567890123456789012345 ABCDEFGHIJABCDEFGHIJABCDEF
I1 26 01234567890123456789012345 ABCDEFGHIJABCDEFGHIJABCDEF
I2 26 01234567890123456789012345 ABCDEFGHIJABCDEFGHIJABCDEF
PK 27 012345678901234567890123456 ABCDEFGHIJABCDEFGHIJABCDEFG
I1 27 012345678901234567890123456 ABCDEFGHIJABCDEFGHIJABCDEFG
I2 27 012345678901234567890123456 ABCDEFGHIJABCDEFGHIJABCDEFG
PK 28 0123456789012345678901234567 ABCDEFGHIJABCDEFGHIJABCDEFGH
I1 28 0123456789012345678901234567 ABCDEFGHIJABCDEFGHIJABCDEFGH
I2 28 0123456789012345678901234567 ABCDEFGHIJABCDEFGHIJABCDEFGH
PK 29 01234567890123456789012345678 ABCDEFGHIJABCDEFGHIJABCDEFGHI
I1 29 01234567890123456789012345678 ABCDEFGHIJABCDEFGHIJABCDEFGHI
I2 29 01234567890123456789012345678 ABCDEFGHIJABCDEFGHIJABCDEFGHI
PK 30 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
I1 30 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
I2 30 012345678901234567890123456789 ABCDEFGHIJABCDEFGHIJABCDEFGHIJ
